### PR TITLE
[canvaskit] Correctly get the directionality of the text boxes

### PR DESCRIPTION
**Pull request by hterkelsen**
The old code assumed the directionality of the TextBoxes was the same as the Paragraph. The new code gets the directionality from CanvasKit (which it computes using Unicode data).

Fixes https://github.com/flutter/flutter/issues/78550

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/web_ui/test/canvaskit/text_test.dart [hunk 1]
```diff
@@ -54,6 +54,21 @@ void testMain() {
         expect(paragraph, isNotNull);
       }
     });
+
+    test('works for https://github.com/flutter/flutter/issues/78550', () {
```
Comment (kevmoo) [inline lib/web_ui/test/canvaskit/text_test.dart:58]: Love linking to the regression, but shouldn't the name of the test be a bit more descriptive?

**Review by kevmoo — APPROVED** [review]

File: lib/web_ui/test/canvaskit/text_test.dart (hunk 1 above)
Comment (hterkelsen) [reply lib/web_ui/test/canvaskit/text_test.dart]: Done

**Comment by hterkelsen** [comment]
I had to add an exception for `canvaskit/text.dart` to use `dart:js_util` because CanvasKit puts the `direction` property as a JS property on the `Float32List`. Raw JavaScript access is the only way to get at it.

File: lib/web_ui/lib/src/engine/canvaskit/text.dart [hunk 2]
```diff
@@ -729,12 +728,16 @@ class CkParagraph extends SkiaObject<SkParagraph> implements ui.Paragraph {
 
     for (int i = 0; i < skRects.length; i++) {
       final Float32List rect = skRects[i];
+      final Object? skTextDirection = getJsProperty(rect, 'direction');
+      final int skTextDirectionValue = skTextDirection != null
+          ? getJsProperty(skTextDirection, 'value')
+          : 0;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/canvaskit/text.dart:734]: Do we need to check for null? My gut feeling is that this is always available.

**Review by yjbanov — APPROVED** [review]
lgtm

File: lib/web_ui/lib/src/engine/canvaskit/text.dart (hunk 2 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/canvaskit/text.dart]: No. Done

**Pull request merged**